sub/filter.c

```diff
@@ -0,0 +1,240 @@
+/*
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdlib.h>
+#include <assert.h>
+#include <string.h>
+#include "common/common.h"
+#include "common/msg.h"
+#include "sd.h"
+
+// Filter for removing subtitle additions for deaf or hard-of-hearing (SDH)
+// This is for English, but may in part work for others too.
+// The intention is that it can be on always so may not remove
+// all SDH parts.
+// This is for filtering ASS encoded subtitles
+//
+// The incoming ass line is overwritten in place (starting at text position)
+// If result is empty line the first postion of the ass line will be set to NULL.
+//
+// Future optimization could be to include state so format and n_ignored
+// could be done once.
+//
+// Future enhancement could be to add "- " when a speaker label
+// is removed. Need to check what is used in current file as some
+// use "-" and some nothing at all.
+void filter_SDH(struct sd *sd, char * format, int n_ignored, char * ass, int filter_harder)
+{
+    char * p = ass;
+
+    if (!format)
+        return;
+
+    int comma = 0;
+    // scan format line to find where the text is
+    for (char *c = format; *c; c++) {
+        if (*c == ',') {
+            comma++;
+            if (strncasecmp(c + 1, "Text", 4) == 0)
+                break;
+        }
+    }
+    // if preprocessed line some commas are skipped
+    comma -= n_ignored;
+
+    // locate text
+    for (int k = 0; k < comma; k++) {
+        while (*p && *p != ',')
+            p++;
+        if (*p == ',')
+            p++;
+    }
+    // p is now at start of text
+    char * ass_text_start = p;
+    char * t = p;   // current writing position
+
+    // go through the lines in the text
+    // they are separated by \N
+    for (;*p;) {
+        // now at beginning of a line
+        char * text_with_ass_formatting_start = t;
+        // skip leading ass formatting like bold or italic
+        while (*p == '{') {
+            *t++ = *p++;
+            for (; *p && *p != '}'; *t++ = *p++);
```

Comment:
This is a good example of C string code that's both dumb and too clever at the same time.
- it has a redundant check (`*p` is either 0 or '}', why check for both?)
- it doesn't use the first expression of the `for`
- it mutates the data in the third expression of the `for`
- it has no `for` body (good example for how to use `for` how to obfuscate)
- there's nothing that guarantees that `t` doesn't go past its allocation, or that `t` doesn't stomp over `p` before it's read